### ISSUE TITLE
fatal: unsafe repository (REPO is owned by someone else) in other workflow steps after running checkout

### ISSUE THREAD

**Issue by thboop** [actions/checkout#766]
### Description
Git recently pushed a change in response to a [cve](https://github.blog/2022-04-12-git-security-vulnerability-announced/) that causes git commands to fail if the parent directory changes ownership from the current directory. You may see information like

```
  /usr/bin/git remote add origin https://github.com/wez/wezterm
  Error: fatal: unsafe repository ('/__w/wezterm/wezterm' is owned by someone else)
```

on self hosted runners, or if your job uses a container.

This was also failing the checkout action(#760), so we shipped a change where when we run the checkout action, we set this config, but we don't persist that configuration.

So, you may still see this error if your job uses git commands outside of the checkout action. If so, you just need to set the configuration value yourself, you can see how to do that below.

### Workaround
In a step if you are running in a job container, or as a part of your container action, please set the `GITHUB_WORKSPACE` as a safe directory. 

```
git config --global --add safe.directory "$GITHUB_WORKSPACE"
```

If your github workspace starts off with `//`, you may need to set it via
```
git config --global --add safe.directory "%(prefix)/$GITHUB_WORKSPACE"
```

### Why is the parent directory owned by a different user?
When the runner maps the working directory mounts into your job container or step container they are owned by the runner user, not the container user, causing this issue. While any folders created may be owned by the container user.

### Whats next

The actions team is exploring solutions for a fix on the ecosystem level for this issue. 

We could try to persist this temporary global configuration we set in checkout for the duration of your job, but there are few problems with that:
1. If you run checkout on the root machine, and you have a container action with git commands, you are still going to fail unless you set the config in that container, which checkout can't do for another step
2. Overwriting the git global config and not persisting any changes back to the original global config may break some user expectations on self hosted runners.
3. It only really addresses this issue for checkout users, but this is more of an actions ecosystem problem

This is better solved at an actions ecosystem level, rather than solving it in the `checkout` action. That way, users not using checkout and users using container actions can take advantage of that solution. This is something our team is actively looking into now.

**Comment by me-and** [comment]
@thboop I agree with everything you've written here, and thank you for getting the fix for this action sorted so quickly! I think I'm a little confused about why this ticket is here, though; should this be raised and tracked over at [actions/runner](https://github.com/actions/runner/issues)?

**Comment by thboop** [comment]
> @thboop I agree with everything you've written here, and thank you for getting the fix for this action sorted so quickly! I think I'm a little confused about why this ticket is here, though; should this be raised and tracked over at [actions/runner](https://github.com/actions/runner/issues)?

🏅 Its a good point that this is more appropriate for the runner repository, however given the nature of this breaking workflows, and the overall concern and questions folks had about what is happening and why, I though I would put it closer to where we are seeing the most users comment about this issue. I may move it in the near future over to the runner repo.

**Comment by JasonGross** [comment]
> Simply set the `GITHUB_WORKSPACE` as a safe directory.
> 
> ```
> git config --global --add safe.directory "$GITHUB_WORKSPACE"
> ```

It seems like this is insufficient if there are submodules? ([log](https://github.com/mit-plv/fiat-crypto/runs/6037359657?check_suite_focus=true))

**Comment by spyoungtech** [comment]
> v3, v3.0.1, v2 and v2.4.1 all contain the fix for this issue.

Will it be possible for a fix be provided for `actions/checkout@v1`?

I am a publisher of a docker-based Action that depends on `actions/checkout` as a prerequisite step. Many of our users have been using this action since 2019 and, as a result, are often using the `v1` version of this action. If a fix would be provided for `v1` it may obviate the need for our users to update the action.

I also understand there may be a fix at the Actions ecosystem level, I'm hopeful would provide the same effect in absence of a fix for `v1`.

In the meantime, we are working on providing guidance to our affected users to deal with this by upgrading to `actions/checkout@v3`, although we are concerned that the version hop from `v1` may introduce unexpected issues for our users.


In any case, it would be good to know if the team is looking at the possibility of porting a fix for `v1`. Thanks!

**Comment by alexprogrammr** [comment]
I have a similar issue running workflow in the container. Probably, that happens because `$GITHUB_WORKSPACE` and `github.workspace` resolves to different paths (https://github.com/actions/checkout/issues/785). Seems like `actions/checkout` adds only `github.workspace` path to the git safe directory configuration.
Probably it makes sense to add `$GITHUB_WORKSPACE` path in `action/checkout` implementation too.

**Comment by trevor92** [comment]
Hello everyone, this issue looks to be related to this: https://github.com/actions/checkout/issues/766

**Comment by Skarlso** [comment]
This popped up on our radar at https://github.com/weaveworks/eksctl after upgrading to v3 AND we updated one of our build containers. With an old container ( and it looks like an older git version ) this IS working. But with a new container this is now breaking and we have to re-apply the safe folder before our flow begins. Checkout applies the safe folder to `/__w/eksctl/eksctl` but the action coming after it, doesn't work anymore. ( fails with not a git directory error ).

```
Run ./.github/actions/setup-identity
Run git config user.name "weaveworksbot"
fatal: not in a git directory
Error: Process completed with exit code [12](https://github.com/weaveworks/eksctl/runs/6919573478?check_suite_focus=true#step:5:14)8.
```

**Comment by rtxa** [comment]
If you use a Docker container, downgrading to `actions/checkout@v1` also fixes the issue. Not sure why isn't mentioned...

**Comment by dave-code-ruiz** [comment]
I fixed it with :

git config --global --add safe.directory "*"

but i dont know if it is the correct way

**Comment by guettli** [comment]
BTW, check that `$HOME` is correct by comparing `getent passwd $(id -un) | cut -d: -f6` with `$HOME`. You might be affected by: [HOME is overridden in Container](https://github.com/actions/runner/issues/863)
